sql: soften AOST check to be a pgerror

otan opened the pull request:
Resolves #82179

Release note (bug fix): Previously, using AS OF SYSTEM TIME of two
different statements in the same line would result in an assertion
error. This is now a PG error with code 0A000.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82504)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -775,9 +775,11 @@ func (ex *connExecutor) handleAOST(ctx context.Context, stmt tree.Statement) err
 		}
-		return errors.AssertionFailedf(
+		return pgerror.Newf(
+			pgcode.FeatureNotSupported,
```
could also be InvalidParameter, but up to you

otan replied:
i'm keeping it as this to be consistent with the other places

rafiss reviewed (approved):
lgtm! do you think there are any cases that should still be treated as an assertion error? if so, then maybe we want to detect this differently.

rafiss commented on pkg/ccl/logictestccl/testdata/logic_test/as_of:
```diff
@@ -29,6 +29,9 @@ BEGIN AS OF SYSTEM TIME follower_read_timestamp(); SELECT * FROM t
 statement ok
 ROLLBACK
 
+statement error pgcode 0A000 cannot specify AS OF SYSTEM TIME with different timestamps
```
ah actually i just realized, this is affecting 22.1, which has `enable_implicit_transaction_for_batch_statements` set to false.

based on the stack trace in https://github.com/cockroachdb/cockroach/issues/82179 it looks like this happened when executing a prepared statement (there is `execPortal` in the stack trace)

also it shows that it's from this line: https://github.com/cockroachdb/cockroach/blob/5b78463ed2e7106a8477b63fa837564ad02bb510/pkg/sql/conn_executor_exec.go#L776

rafiss replied:
oh that's the line you updated. ok. i'm fine with this change

rafiss replied:
i think it's because we also use an implicit transaction for the extended protocol now (https://github.com/cockroachdb/cockroach/pull/76792), so a test for that  too would be nice.

otan replied:
do you know how it'd be triggered using the extended protocol?

rafiss replied:
yeah, i think  if you do something like:

```
Parse SELECT .. AOST '-1us`
Bind
Exec
Parse SELECT ... AOST '-4us'
Bind
Exec
Sync
```

hmm but that wouldn't cause it during ExecPortal

what if you do
```
Parse SELECT .. AOST follower_read_timestamp()
Bind
Exec
Sync
```

otan replied:
dones

otan committed:
sql: soften AOST check to be a pgerror

Release note (bug fix): Previously, using AS OF SYSTEM TIME of two
different statements in the same line would result in an assertion
error. This is now a PG error with code 0A000.

otan commented:
bors r=rafiss
